### PR TITLE
Implement conversions for non-string keyed maps

### PR DESCRIPTION
**Pull request by dfaust**
Currently only maps with keys of type `String` can be used. This PR adds a macro to automatically generate the `AsRust` trait for many different types of hash maps. It also replaces the implementation of `From<HashMap<String, T>>` for `Bytes` with the more general form `From<HashMap<K, V>>`.

### CODE REVIEW

File: src/types/map.rs [hunk 1]
```diff
@@ -12,472 +12,332 @@ use error::Result;
...
-                        Ok(self.data
-                            .iter()
-                            .fold(map, |mut acc, (k, vb)| {
-                                acc.insert(k.clone(), decode_blob(vb.as_plain()).unwrap());
-                                return acc;
-                            }))
-                    }
-                    _ => unreachable!(),
-                }
+macro_rules! data_as_rust {
```
Comment (AlexPikalov) [inline src/types/map.rs]: It looks as a macros which could be used across the whole `types` module in order to reduce boilerplate in a same fashion as it had been done here. So could you please move this macros into some separate module which we could share inside `types`.

Comment (AlexPikalov) [inline src/types/map.rs]: Probably we could do it in a separate PR as well.

Comment (dfaust) [inline src/types/map.rs]: How about renaming `data_as_rust` to `decode_type` and moving it to `data_serialization_types.rs`?

Comment (AlexPikalov) [inline src/types/map.rs]: @dfaust That also sounds good for me.

Comment (AlexPikalov) [inline src/types/map.rs]: BTW `as_rust` is also a good name for this macros. As for primitive types it does pretty much the same as `AsRust` trait is supposed to do.

**Comment by harrydevnull** [comment]
@dfaust @alex would this right time to change unwraps and unreachable to result ?

**Comment by harrydevnull** [comment]
Sorry for grammar and typos on phone !!!

**Comment by AlexPikalov** [comment]
@harrydevnull I believe we could do it in a separate request. After that update it would be much easier to revise all that stuff.
